src/canisters/frontend/ic-certified-assets/src/lib.rs

```diff
@@ -24,15 +29,55 @@ thread_local! {
 
 #[update]
 #[candid_method(update)]
-fn authorize(other: Principal) {
+async fn authorize(other: Principal) {
+    let caller = caller();
+    let is_authorized = STATE.with(|s| s.borrow().is_authorized(&caller));
+    if is_authorized {
+        STATE.with(|s| {
+            if let Err(msg) = s.borrow_mut().authorize(&caller, other) {
+                trap(&msg);
+            }
+        })
+    } else {
+        match canister_status(CanisterIdRecord {
+            canister_id: ic_cdk::api::id(),
+        })
+        .await
+        {
+            Err((code, msg)) => trap(&format!(
+                "Caller is not authorized. Failed to determine if caller is canister controller with code {:?} and message '{}'",
+                code, msg
+            )),
+            Ok((a,)) => {
+                if a.settings.controllers.contains(&caller) {
+                    STATE.with(|s| s.borrow_mut().authorize_unconditionally(other))
+                } else {
+                    trap(
+                        "The caller is not authorized and not part of the canister's controllers.",
+                    );
+                }
+            }
+        }
+    }
+}
+
+#[update(guard = "is_authorized")]
```

Comment:
I think if `authorize` allows controllers that are not on the authorized list, then so should `deauthorize`.  Thinking about this sequence of events from the perspective of such a controller:
1. Controller calls 'authorize' for a given principal
2. Controller for whatever reason wants to deauthorize that same principal (this could be immediately, or theoretically after some time, perhaps due to a proposal passing)
3. Controller calls `deauthorize` for that same principal
4. Access denied. Controller would have to add themselves to the authorized list first.